File: sdk/src/aggregate.ts
// Phase 7.2 — federation aggregation primitive.
//
// L'agent SDK utilise cette utility pour découvrir les oracles SatRank-
// compatible et filtrer ceux dont la calibration history meet ses
// critères. Pure helper — pas wired dans fulfill() automatiquement (c'est
// l'agent qui choisit quand et comment fédérer).
//
// Usage typique :
//
//   import { fetchOraclePeers, filterByCalibrationError } from '@satrank/sdk/aggregate';
//
//   const peers = await fetchOraclePeers({ baseUrl: 'https://satrank.dev' });
//   const trusted = filterByCalibrationError(peers, {
//     maxStaleSec: 7 * 86400,
//     minCatalogueSize: 50,
//   });
//   // trusted = liste de OraclePeer ayant publié il y a < 7d et catalogue ≥ 50
//   // L'agent peut alors interroger chacun via /api/intent ou via leur DVM
//   // kind 5900 et aggréger les réponses (Bayesian model averaging — à
//   // implémenter quand on aura des kind 30783 calibrations cross-oracle).

export interface OraclePeer {
  oracle_pubkey: string;
  lnd_pubkey: string | null;
  catalogue_size: number;
  calibration_event_id: string | null;
  last_assertion_event_id: string | null;
  contact: string | null;
  onboarding_url: string | null;
  last_seen: number;
  first_seen: number;
  age_sec: number;
  stale_sec: number;
  latest_announcement_event_id: string | null;
}

export interface FetchOraclePeersOptions {
  /** Base URL d'un oracle SatRank-compatible. Default 'https://satrank.dev'. */
  baseUrl?: string;
  /** Limit max de peers à récupérer (server clamp à 200). */
  limit?: number;
  /** fetch impl injectable pour tests. */
  fetchImpl?: typeof fetch;
}

export interface FetchOraclePeersResult {
  peers: OraclePeer[];
  count: number;
  source_oracle: string;
}

/** Récupère la liste des peers découverts par un oracle SatRank donné via
 *  GET /api/oracle/peers. */
export async function fetchOraclePeers(opts: FetchOraclePeersOptions = {}): Promise<FetchOraclePeersResult> {
  const baseUrl = opts.baseUrl ?? 'https://satrank.dev';
  const limit = opts.limit ?? 50;
  const fetchImpl = opts.fetchImpl ?? fetch;
  const url = new URL('/api/oracle/peers', baseUrl);
  url.searchParams.set('limit', String(limit));
  const resp = await fetchImpl(url.toString(), {
    method: 'GET',
    headers: { Accept: 'application/json' },
  });
  if (!resp.ok) {
    throw new Error(`fetchOraclePeers: ${resp.status} ${resp.statusText}`);
  }
  const body = await resp.json() as { data?: { peers?: OraclePeer[]; count?: number } };
  const peers = body.data?.peers ?? [];
  return {
    peers,
    count: body.data?.count ?? peers.length,
    source_oracle: baseUrl,
  };
}

export interface FilterPeersOptions {
  /** Rejeter les peers qui n'ont pas re-publié depuis > N secondes. Default 7 jours. */
  maxStaleSec?: number;
  /** Catalogue size minimum. Default 50. */
  minCatalogueSize?: number;
  /** Exiger un calibration_event_id présent (= peer publie son calibration history).
   *  Default true. */
  requireCalibration?: boolean;
  /** Age minimum du peer (first_seen). Pas de Sybil-resistance forte ici, juste
   *  un filter age basique. Default 0 (no lower bound). */
  minAgeSec?: number;
}

/** Filtre une liste de peers selon les critères trust de l'agent. Pure
 *  function. */
export function filterByCalibrationError(
  peers: OraclePeer[],
  opts: FilterPeersOptions = {},
): OraclePeer[] {
  const maxStale = opts.maxStaleSec ?? 7 * 86400;
  const minCat = opts.minCatalogueSize ?? 50;
  const requireCalib = opts.requireCalibration ?? true;
  const minAge = opts.minAgeSec ?? 0;
  return peers.filter((p) => {
    if (p.stale_sec > maxStale) return false;
    if (p.catalogue_size < minCat) return false;
    if (requireCalib && !p.calibration_event_id) return false;
    if (p.age_sec < minAge) return false;
    return true;
  });
}

export interface AggregateOraclesOptions extends FetchOraclePeersOptions, FilterPeersOptions {}

/** Helper combiné : fetch + filter + return liste utilisable pour
 *  cross-oracle queries. L'agent peut ensuite interroger chaque peer
 *  via /api/intent ou son DVM. */
export async function aggregateOracles(opts: AggregateOraclesOptions = {}): Promise<{
  peers: OraclePeer[];
  total_discovered: number;
  trusted_count: number;
  source_oracle: string;
}> {
  const fetched = await fetchOraclePeers(opts);
  const trusted = filterByCalibrationError(fetched.peers, opts);
  return {
    peers: trusted,
    total_discovered: fetched.count,
    trusted_count: trusted.length,
    source_oracle: fetched.source_oracle,
  };
}

File: python-sdk/satrank/aggregate.py
"""Phase 7.2 — federation aggregation primitive.

The agent SDK uses this utility to discover SatRank-compatible oracles and
filter the ones whose calibration history meets its trust criteria. Pure
helper — not auto-wired into ``fulfill()`` (the agent decides when and how
to federate).

Mirrors @satrank/sdk's ``aggregate.ts``. See the TS file for the original
docstring.

Typical usage::

    from satrank.aggregate import fetch_oracle_peers, aggregate_oracles

    # 1) Discover peers from one seed oracle:
    fetched = await fetch_oracle_peers(base_url="https://satrank.dev")

    # 2) Filter by your own trust criteria:
    result = await aggregate_oracles(
        base_url="https://satrank.dev",
        max_stale_sec=7 * 86400,
        min_catalogue_size=50,
        require_calibration=True,
    )
    # result["peers"] = list of OraclePeer matching the criteria.
    # The agent then queries each via /api/intent or its DVM.
"""

from __future__ import annotations

from typing import Any

import httpx

from satrank.types import OraclePeer


async def fetch_oracle_peers(
    *,
    base_url: str = "https://satrank.dev",
    limit: int = 50,
    http: httpx.AsyncClient | None = None,
    timeout_s: float = 15.0,
) -> dict[str, Any]:
    """Fetch the peer list of a SatRank-compatible oracle.

    Returns a dict::

        {
            "peers": list[OraclePeer],
            "count": int,
            "source_oracle": str,
        }

    Server-side ``limit`` clamps to 200.
    """
    url = f"{base_url.rstrip('/')}/api/oracle/peers"
    params = {"limit": str(limit)}
    own_client = http is None
    client = http or httpx.AsyncClient(timeout=timeout_s)
    try:
        resp = await client.get(
            url, params=params, headers={"Accept": "application/json"}
        )
        resp.raise_for_status()
        body = resp.json()
    finally:
        if own_client:
            await client.aclose()
    data = body.get("data") or {}
    peers: list[OraclePeer] = list(data.get("peers") or [])
    count = int(data.get("count") or len(peers))
    return {"peers": peers, "count": count, "source_oracle": base_url}


def filter_by_calibration_error(
    peers: list[OraclePeer],
    *,
    max_stale_sec: int = 7 * 86400,
    min_catalogue_size: int = 50,
    require_calibration: bool = True,
    min_age_sec: int = 0,
) -> list[OraclePeer]:
    """Filter a peer list against the agent's trust criteria. Pure function.

    Defaults match the TypeScript ``filterByCalibrationError``.
    """
    out: list[OraclePeer] = []
    for p in peers:
        if int(p.get("stale_sec", 0)) > max_stale_sec:
            continue
        if int(p.get("catalogue_size", 0)) < min_catalogue_size:
            continue
        if require_calibration and not p.get("calibration_event_id"):
            continue
        if int(p.get("age_sec", 0)) < min_age_sec:
            continue
        out.append(p)
    return out


async def aggregate_oracles(
    *,
    base_url: str = "https://satrank.dev",
    limit: int = 50,
    max_stale_sec: int = 7 * 86400,
    min_catalogue_size: int = 50,
    require_calibration: bool = True,
    min_age_sec: int = 0,
    http: httpx.AsyncClient | None = None,
    timeout_s: float = 15.0,
) -> dict[str, Any]:
    """Combined helper: fetch peers from a seed oracle, then filter them
    by the agent's trust criteria.

    Returns::

        {
            "peers": list[OraclePeer],     # peers that passed the filter
            "total_discovered": int,        # count from the seed oracle
            "trusted_count": int,           # len(peers)
            "source_oracle": str,           # the seed URL used
        }

    The agent can then iterate ``peers`` and query each via /api/intent or
    its DVM (kind 5900) for cross-oracle Bayesian model averaging.
    """
    fetched = await fetch_oracle_peers(
        base_url=base_url, limit=limit, http=http, timeout_s=timeout_s
    )
    trusted = filter_by_calibration_error(
        fetched["peers"],
        max_stale_sec=max_stale_sec,
        min_catalogue_size=min_catalogue_size,
        require_calibration=require_calibration,
        min_age_sec=min_age_sec,
    )
    return {
        "peers": trusted,
        "total_discovered": fetched["count"],
        "trusted_count": len(trusted),
        "source_oracle": fetched["source_oracle"],
    }
